zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,195 @@
+
+use crate::errors::VMError;
+use crate::signature::multikey::Multikey;
+use crate::signature::musig::*;
+use crate::signature::VerificationKey;
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use rand;
+use subtle::ConstantTimeEq;
+
+#[derive(Copy, Clone, Debug)]
+pub struct Nonce(Scalar);
+
+#[derive(Copy, Clone)]
+pub struct NoncePrecommitment([u8; 32]);
+
+#[derive(Copy, Clone, Debug)]
+pub struct NonceCommitment(RistrettoPoint);
+
+#[derive(Copy, Clone, Debug)]
+pub struct Siglet(Scalar);
+
+pub struct PartyAwaitingPrecommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: PrivKey,
+    r_i: Nonce,
+    R_i: NonceCommitment,
+}
+
+pub struct PartyAwaitingCommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: PrivKey,
+    r_i: Nonce,
+    nonce_precommitments: Vec<NoncePrecommitment>,
+}
+
+pub struct PartyAwaitingSiglets {
+    multikey: Multikey,
+    c: Scalar,
+    nonce_commitments: Vec<NonceCommitment>,
+}
+
+impl NonceCommitment {
+    fn precommit(&self) -> NoncePrecommitment {
+        let mut h = Transcript::new(b"MuSig.nonce-precommit");
+        h.commit_point(b"R_i", &self.0.compress());
+        let mut precommitment = [0u8; 32];
+        h.challenge_bytes(b"precommitment", &mut precommitment);
+        NoncePrecommitment(precommitment)
+    }
+}
+
+impl PartyAwaitingPrecommitments {
+    pub fn new(
+        // The message `m` should already have been fed into the transcript
+        transcript: &Transcript,
+        x_i: PrivKey,
+        multikey: Multikey,
+    ) -> (Self, NoncePrecommitment) {
+        let mut rng_transcript = transcript.clone();
+        rng_transcript.commit_scalar(b"x_i", &x_i.0);
+        let mut rng = rng_transcript.build_rng().finalize(&mut rand::thread_rng());
+
+        // Generate ephemeral keypair (r_i, R_i). r_i is a random nonce.
+        let r_i = Nonce(Scalar::random(&mut rng));
+        // R_i = generator * r_i
+        let R_i = NonceCommitment(RISTRETTO_BASEPOINT_POINT * r_i.0);
+        // Make H(R_i)
+        let precommitment = R_i.precommit();
+
+        (
+            PartyAwaitingPrecommitments {
+                transcript: transcript.clone(),
+                multikey,
+                x_i,
+                r_i,
+                R_i,
+            },
+            precommitment,
+        )
+    }
+
+    pub fn receive_precommitments(
+        self,
+        nonce_precommitments: Vec<NoncePrecommitment>,
+    ) -> (PartyAwaitingCommitments, NonceCommitment) {
+        // Store received nonce precommitments in next state
+        (
+            PartyAwaitingCommitments {
+                transcript: self.transcript,
+                multikey: self.multikey,
+                x_i: self.x_i,
+                r_i: self.r_i,
+                nonce_precommitments,
```

Comment:
if you'd like you can do `PartyAwaitingCommitments { nonce_precommitments, .. self }` to fill in the rest of the fields from `self`.

Reply:
I think this only works if `self` is of the same type as the struct you're filling in? 

I tried this, and I got the error "mismatched types... expected struct `PartyAwaitingCommitments`, found type `PartyAwaitingPrecommitments`.

I tried to look up documentation on filling the rest of the fields using `.. self` but couldn't find any, I'm curious where to find more information on using this trick?

Reply:
Yeah, maybe it only works with the same type, oops.
https://doc.rust-lang.org/1.6.0/book/structs.html#update-syntax

<img width="713" alt="image" src="https://user-images.githubusercontent.com/698/54331778-c01eb080-45d8-11e9-8d27-2b5b123d97c8.png">

Reply:
oh interesting!
yeah that makes sense that it would only work with the same type, because then the compiler could know what fields are supposed to correspond with each other, I think? Anyway, I learned something new!